docs/source/user/stencil.rst

```diff
@@ -0,0 +1,205 @@
+          [  0.,  81.,  82.,  83.,  84.,  85.,  86.,  87.,  88.,   0.],
+          [  0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.]])
+   >>> input_arr.dtype
+   dtype('int64')
+   >>> output_arr.dtype
+   dtype('float64')
+
+Note that the stencil decorator has determined that the output type
+of the specified stencil kernel is ``float`` and has thus created the
+output array as ``float`` while the input array is of type ``int``.
+
+Stencil Parameters
+==================
+
+Stencil kernel definitions may take any number of arguments with
+the following provisions.  The first parameter must be an array.
+The size and shape of the output array will be the same as the
+first parameter.  Additional parameters may either be scalars or
+arrays.  For array parameters, those arrays must be at least as large
+as the first array in each dimension.  Array indexing is relative for
+all such input array parameters.
+
+.. _stencil-kernel-shape-inference:
+
+Kernel shape inference and border handling
+==========================================
+
+In the above example and in most cases, the array indexing in the 
+stencil kernel will exclusively use Integer literals.
+In such cases, the stencil decorator is able to analyze the stencil
+kernel to determine its size.  In the above example, the stencil
+decorator determines that the kernel is 3x3 since indices -1 to 1
+are used for both the first and second dimensions.  Note that
+the stencil decorator also correctly handles non-symmetric and 
+non-square stencil kernels.
+
+Based on the size of the stencil kernel, the stencil decorator is
+able to compute the size of the border in the output array.  If
+applying the kernel to some element of input array would cause
+an index to be out-of-bounds then that element belongs to the border
+of the output array.  In the above example, points -1 and +1 are
+accessed in each dimension and thus the output array has a border
+of size one in all dimensions.
+
+
+Stencil decorator options
+=========================
+
+While the stencil decorator may be augmented in the future to 
+provide additional mechanisms for border handling, at the moment
+the stencil decorator currently supports only one option.
+
+.. _stencil-neighborhood:
+
+``neighborhood``
+----------------
+
+Sometimes it may be inconvenient to write the stencil kernel
+exclusively with Integer literals.  For example, let us say we
+would like to compute the trailing 30-day moving average of a
+time series of data.  One could write 
+``(a[-29] + a[-28] + ... + a[-1] + a[0]) / 30`` but the stencil
+decorator offers a more concise form using the ``neighborhood``
+option::
+
+   @stencil(neighborhood=((-29,0),))
+   def kernel2(a):
+       cumul = 0
+       for i in range(-29,1):
+           cumul += a[i]
+       return cumul / 30
+
+The neighborhood option is a tuple of tuples.  The outer tuple's
+length is equal to the number of dimensions of the input array.
```

Comment:
Practically, at run time if these conditions are not met can a suitable be error raised with a message explaining? Just tried this:
```python
from numba import stencil
import numpy as np

@stencil(neighborhood=((-29,0),))
def k(a):
    cumul = 0
    for i in range(-29, 1):
        cumul += a[i]
    return cumul / 30

input_arr = np.arange(300).reshape((3, 100))
print(input_arr)
output_arr = k(input_arr)
print(output_arr)
```
which raises a typing error as type unification fails. The other option is of course to support the incredibly complicated broadcasting :)

Reply:
I've fixed this by throwing a RuntimeError.  Is that the right error to throw you think?

Reply:
I'd probably go with a `ValueError` as it derives from an incorrectly specified value of a correct type.

Reply:
Done.